main/drivers/CANDriver.c

```diff
@@ -0,0 +1,293 @@
+#include "CANDriver.h"
+
+#include <esp_log.h>
+#include <string.h>
+
+#define CAN_STACK_SIZE 2048 // Minimum
+
+static const char *TAG = "CANDriver";
+
+static CANDriver_t driver;
+
+/**
+ * The ISR handler for the INT1 pin of device triggered on rising edge
+ * @param arg The can bus number
+ */
+void IRAM_ATTR CANDriverISRHandler(void *arg) {
+    uint32_t bus = (uint32_t)arg;
+
+    // set the corrsponding event to chip with data (this basically defers
+    // the SPI reading to a later time that is not in an interrupt).
+    xEventGroupSetBitsFromISR(driver.messageEvents, 1 << bus, pdTRUE);
+}
+
+/**
+ * The FreeRTOS task for reading
+ * @param arg UNUSED (part to FreeRTOS spec)
+ */
+static void CANDriverTask(void *arg) {
+    EventBits_t activeEvents = 0;
+
+    // wait until the mask has atleast on bit set
+    while (!driver.deviceEventMask)
+        vTaskDelay(1);
+
+    for (;;) {
+        // determine which bus has just got data
+        if (activeEvents) {
+            // get any additional events while there are still messages in the FIFO
+            activeEvents |= xEventGroupClearBits(driver.messageEvents, driver.deviceEventMask);
+        } else {
+            // get any evens, wait until an even happens or 10ms without event
+            activeEvents |=
+                xEventGroupWaitBits(driver.messageEvents, driver.deviceEventMask, true, false, portMAX_DELAY);
+        }
+
+        // read messages from the CAN buses that currently have messages
+        for (int i = 0; activeEvents >> i && i <= (0x1 << CAN_BUS_COUNT); i++) {
```

Comment:
I think this loop would be a little clearer if you did something like this:
```suggestion
        for (int i = 0; activeEvents != 0 && i <= (0x1 << CAN_BUS_COUNT); i++, activeEvents >> i) {
```

Reply:
Might also be faster if you swap the checks because I imagine the left shift will fail more often

Reply:
This suggestion would not work since it would destroy the state of activeEvents. This state needs to persist cross multiple iterations.

Reply:
I could move the activeEvents check to a separate if statement that would break out of the for loop, but I figured this was more explicit that we read from each of the CAN chips as long as there are messages.